Symfony/CS/Tests/AbstractIntegrationTest.php

```diff
@@ -0,0 +1,271 @@
+<?php
+
+/*
+ * This file is part of the PHP CS utility.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Symfony\CS\Tests;
+
+use Symfony\Component\Filesystem\Exception\IOException;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\CS\ErrorsManager;
+use Symfony\CS\FileCacheManager;
+use Symfony\CS\Fixer;
+use Symfony\CS\FixerInterface;
+
+/**
+ * Integration test base class.
+ *
+ * This test searches for '.test' fixture files in the given directory.
+ * Each fixture file will be parsed and tested against the expected result.
+ *
+ * Fixture files have the following format:
+ *
+ * --TEST--
+ * Example test description
+ * --CONFIG--
+ * level=symfony|none|psr0|psr1|psr2|symfony
+ * fixers=fixer1,fixer2,...*
+ * --fixers=fixer3,fixer4,...**
+ * --INPUT--
+ * Code to fix
+ * --EXPECT--
+ * Expected code after fixing***
+ *
+ *   * Additional fixers may be omitted.
+ *  ** Black listed filters may be omitted.
+ * *** When the expected block is omitted the input is expected not to
+ *     be changed by the fixers.
+ *
+ * @author SpacePossum
+ *
+ * @internal
+ */
+abstract class AbstractIntegrationTest extends \PHPUnit_Framework_TestCase
+{
+    private static $builtInFixers;
+
+    public static function setUpBeforeClass()
+    {
+        $tmpFile = static::getTempFile();
+        if (!is_file($tmpFile)) {
+            $dir = dirname($tmpFile);
+            if (!is_dir($dir)) {
+                $fs = new Filesystem();
+                $fs->mkdir($dir, 0766);
+            }
+        }
+    }
+
+    public static function tearDownAfterClass()
+    {
+        unlink(static::getTempFile());
+    }
+
+    /**
+     * @dataProvider getTests
+     *
+     * @see doTestIntegration()
+     */
+    public function testIntegration($testFileName, $testTitle, $fixers, $input, $expected = null)
+    {
+        $this->doTestIntegration($testFileName, $testTitle, $fixers, $input, $expected);
+    }
+
+    /**
+     * Creates test data by parsing '.test' files.
+     *
+     * @return array
+     */
+    public function getTests()
+    {
+        $fixturesDir = realpath($this->getFixturesDir());
+        if (!is_dir($fixturesDir)) {
+            throw new \UnexpectedValueException(sprintf('Given fixture dir "%s" is not a directory.', $fixturesDir));
+        }
+
+        $tests = array();
+
+        foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($fixturesDir), \RecursiveIteratorIterator::LEAVES_ONLY) as $file) {
+            if (!preg_match('/\.test$/', $file)) {
+                continue;
+            }
+
+            $test = file_get_contents($file->getRealpath());
+            $fileName = $file->getFileName();
+            if (!preg_match('/--TEST--[\r\n](.*?)\s--CONFIG--[\r\n](.*?)\s--INPUT--[\r\n](.*?[\r\n]*)(?:[\r\n]--EXPECT--\s(.*)|$)/s', $test, $match)) {
```

Comment:
>And still don't know any dev that use \r\n in PHP code (even on Windows) or allow git to mess up with his files and change line ending.

I know of at least 5 developers/maintainers that do this (near me). The thing is IDEs like PHPStorm have a flag to not show white space changes. So as long as everyone on the team switches that one on (which is promoted) and doesn't look at the real changes (or github) and uses no PRs but just pushes to master; there is no problem at all ;)
(same for file permissions which Windows tends to change by default as well..)

back on topic: I'll remove the '\r' : )